src/org/opensolaris/opengrok/util/OptionParser.java

```diff
@@ -0,0 +1,701 @@
+                if (opt.action != null) {
+                    opt.action.accept(value); // 'do' assigned action
+                }
+                optind = ++ii;
+            }
+        }
+        
+        // Prepare to gather any remaining arguments
+        // to send back to calling program.
+        
+        String[] remainingArgs = null;
+        
+        if (optind == -1) {
+            remainingArgs = args;
+        } else if (optind < args.length) {
+            remainingArgs = Arrays.copyOfRange(args, optind, args.length);
+        }
+        
+        return remainingArgs;
+    }
+    
+    private String program() {
+        // Locate main program name from stack trace.
+        StackTraceElement[] stack = Thread.currentThread ().getStackTrace ();
+        StackTraceElement main = stack[stack.length - 1];
+        String[] name = main.getClassName().split("\\.");
+        return name[name.length-1];
+    }
+    
+    private String getPrologue() {
+        // Assign default prologue statement when none given.
+        if (prologue == null) {
+            prologue = "Usage: " + program() + " [options]";
+        }
+        
+        return prologue;
+    }
+    
+    /**
+     * Place text in option summary
+     * @param text to be inserted into option summary.
+     * 
+     * Example usage:
+     *  OptionParser opts = OptionParser.Do( parser -&gt; {
+     *   
+     *    parser.prologue = String.format("Usage: %s [options] bubba smith", program);
+     *    parser.separator("");
+     * 
+     *    parser.on("-y value", "--why me", "This is a description").Do( v -&gt; {
+     *        System.out.println("got " + v);
+     *    });
+     *       
+     *    parser.separator("  ----------------------------------------------");
+     *    parser.separator("  Common Options:");
+     *    ...
+     * 
+     *    parser.separator("  ----------------------------------------------");
+     *    parser.epilogue = "  That's all Folks!";
+     */
+    public void separator(String text) {
+        usageSummary.add(text);
+    }
+    
+    /**
+     * Obtain option summary
+     * @return option summary
+     */
+    public String getUsage() {
+        
+        StringWriter wrt = new StringWriter();
+        String indent = "  ";
+        try (PrintWriter out = new PrintWriter(wrt)) {
+            out.println(getPrologue());
+            for (Object o : usageSummary) {
+                // Need to be able to handle separator strings
+                if (o instanceof String) {
+                    out.println((String)o);
+                } else {
+                    out.println(indent + ((Option)o).getUsage());
+                }
+            }
+            if (epilogue != null) {
+                out.println(epilogue);
+            }
+            out.flush();
+        }
+        return wrt.toString();
+    }
+    
+    /**
+     * Print out option summary.
+     */
+    public void help() {
+        System.out.println(getUsage());
```

Comment:
same here w.r.t. where the output should be printed

Reply:
Again, here do you want me to state in the documentation that the output is going to stdout? (Would you prefer that it goes to stderr?  It doesn't matter to me one way or the other.)

Reply:
I'd like to have the flexibility to choose where it will go.

Reply:
but Vlad, you do have the flexibility....
From the calling program say, System.err.println(parser.getUsage());
I just added the 'help' method as a convenience mechanism.

Reply:
maybe just add an overloaded method help(stream) and use the help() to call it with System.out. I think that should be enogh

Reply:
So do you want this stream to be a  PrintStream, or something more general?

Reply:
`PrintStream` is fine with me